File: ja3.lua
local function split(separator, str)
    local cur, result
    result = {}
    cur = 0
    if (#str == 1) then
        return { str }
    end
    while true do
        next = string.find(str, separator, cur, true)
        if next ~= nil then
            table.insert(result, string.sub(str, cur, next - 1))
            cur = next + 1
        else
            table.insert(result, string.sub(str, cur))
            break
        end
    end
    return result
end

local function hex_to_decimal(hex)
    if #hex ~= 4
    then
        error("hex value '" .. hex .. "' should have 4 letters but have " .. #hex)
    end
    return tostring(tonumber(hex, 16))
end

local grace_codes = {}
grace_codes["0a0a"] = true
grace_codes["1a1a"] = true
grace_codes["2a2a"] = true
grace_codes["3a3a"] = true
grace_codes["4a4a"] = true
grace_codes["5a5a"] = true
grace_codes["6a6a"] = true
grace_codes["7a7a"] = true
grace_codes["8a8a"] = true
grace_codes["9a9a"] = true
grace_codes["aaaa"] = true
grace_codes["baba"] = true
grace_codes["caca"] = true
grace_codes["dada"] = true
grace_codes["eaea"] = true
grace_codes["fafa"] = true


function ja3_concat_decimals(decimals_array)
    local count, result_str
    count = #decimals_array
    result_str = ""
    for k, cipher in pairs(decimals_array) do
        if cipher ~= "" and not grace_codes[cipher] then
            result_str = result_str .. hex_to_decimal(cipher)
            if k ~= count then
                result_str = result_str .. "-"
            end
        end
    end
    return result_str
end

function ja3_curator2sailforce(curator_str)
    local parts, result_str, ciphers

    parts = split(",", curator_str)
    if #parts ~= 5 then
        error("ja3 should have 5 parts")
    end

    if #parts[1] ~= 4 then
        error("ja3 first part should have 4 numbers")
    end

    result_str = hex_to_decimal(parts[1]) .. ","
    result_str = result_str .. ja3_concat_decimals(split(":", parts[2])) .. ","
    result_str = result_str .. ja3_concat_decimals(split(":", parts[3])) .. ","
    result_str = result_str .. ja3_concat_decimals(split(":", parts[4])) .. ","
    result_str = result_str .. ja3_concat_decimals(split(":", parts[5]))

    if have_changes then
        result_str = text:sub(1, -2)
    end

    return result_str
end

--_str = "0304,6a6a:1301:1302:1303:c02c:c02b:cca9:c030:c02f:cca8:c00a:c009:c014:c013,0000:0017:ff01:000a:000b:0010:0005:000d:0012:0033:002d:002b:001b:0015,dada:001d:0017:0018:0019,"
--print(ja3_curator2sailforce(_str))

File: njs/ja3.js
const crypto = require('crypto')

const JA3_HEADER_NAME = "X-Qrator-Cl-Ssl-Params"
const GREASE_TABLE = {
    "0a0a": true, "1a1a": true, "2a2a": true, "3a3a": true,
    "4a4a": true, "5a5a": true, "6a6a": true, "7a7a": true,
    "8a8a": true, "9a9a": true, "aaaa": true, "baba": true,
    "caca": true, "dada": true, "eaea": true, "fafa": true
}

function hex2decimal(hex) {
    let result = parseInt(hex, 16).toString()
    if (result === "NaN") {
        throw new Error("hex2decimal return Nan for value: '" + hex + "'")
    }
    return result
}

function transform_segment(segment) {
    if (segment === "") {
        return ""
    }
    let hex_chunks = segment.split(":")
    let dec_chunks = []

    for (let i in hex_chunks) {
        let hex = hex_chunks[i]
        if (GREASE_TABLE[hex]) {
            continue
        }
        dec_chunks.push(hex2decimal(hex))
    }
    return dec_chunks.join("-")
}

function ja3_curator2sailforce(curator_str) {
    let segments = curator_str.split(",")
    if (segments.length !== 5) {
        throw new Error("ja3 should have 5 segments")
    }
    if (segments[0].length !== 4) {
        throw new Error("ja3 first segment should have 4 digest, but value: " + segments[0])
    }
    let result = []
    for (let i in segments) {
        result.push(transform_segment(segments[i]))
    }

    return result.join(",")
}


function ja3_njs(r, is_digest) {
    let header = r.headersIn[JA3_HEADER_NAME]
    if (!header) {
        return ""
    }
    let ja3_long = ja3_curator2sailforce(header)
    if (!is_digest) {
        return ja3_long
    }
    return crypto.createHash("md5").update(ja3_long).digest("hex")
}

function ja3_njs_digest(r) {
    return ja3_njs(r, true)
}

function ja3_njs_str(r) {
    return ja3_njs(r, false)
}

export default {ja3_curator2sailforce, ja3_njs_digest, ja3_njs_str}

// let _str = "0304,6a6a:1301:1302:1303:c02c:c02b:cca9:c030:c02f:cca8:c00a:c009:c014:c013,0000:0017:ff01:000a:000b:0010:0005:000d:0012:0033:002d:002b:001b:0015,dada:001d:0017:0018:0019,"
// console.log(ja3_curator2sailforce(_str))
